lib/streamlit/commands/navigation.py

```diff
@@ -0,0 +1,189 @@
+
+from streamlit.errors import StreamlitAPIException
+from streamlit.navigation.page import StreamlitPage
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+from streamlit.proto.Navigation_pb2 import Navigation as NavigationProto
+from streamlit.runtime.metrics_util import gather_metrics
+from streamlit.runtime.scriptrunner.script_run_context import (
+    ScriptRunContext,
+    get_script_run_ctx,
+)
+from streamlit.source_util import PageHash, PageInfo
+
+SectionHeader: TypeAlias = str
+
+
+def pages_from_nav_sections(
+    nav_sections: dict[SectionHeader, list[StreamlitPage]]
+) -> list[StreamlitPage]:
+    page_list = []
+    for pages in nav_sections.values():
+        for page in pages:
+            page_list.append(page)
+
+    return page_list
+
+
+def send_page_not_found(ctx: ScriptRunContext):
+    msg = ForwardMsg()
+    msg.page_not_found.page_name = ""
+    ctx.enqueue(msg)
+
+
+@gather_metrics("navigation")
+def navigation(
+    pages: list[StreamlitPage] | dict[SectionHeader, list[StreamlitPage]],
+    *,
+    position: Literal["sidebar"] | Literal["hidden"] = "sidebar",
+) -> StreamlitPage | None:
+    """
+    Configure the available pages in a multipage app.
+
+    Call `st.navigation` in your main script with one or more pages defined by
+    `st.Page`. `st.navigation` returns the current page which can be executed
+    using `Page.run()`.
+
+    When using `st.navigation`, the main script is executed on every app rerun,
+    with the current page executed in-line using `Page.run()`. The set of
+    available pages can be updated with each rerun for dynamic navigation. In
+    this mode, the `pages/` folder is ignored.
+
+    By default, `st.navigation` draws the available pages in the side
+    navigation. This behavior can be changed using the `position=` keyword
+    argument.
+
+    Parameters
+    ----------
+    pages: list[st.Page] or dict[str, list[st.Page]]
+        A list of `st.Page` objects or a dictionary where the keys are section
+        headers and the values are lists of `st.Page` objects.
+
+    position: str
+        The position of the navigation menu. Can be "sidebar" or "hidden".
+
+    Example
+    -------
+    >>> import streamlit as st
+    >>> from pages import page1, page2
+    >>>
+    >>>	pg = st.navigation([st.Page(page1), st.Page(page2)])
+    >>>
+    >>>	st.title("My Awesome App")
+    >>> pg.run()
+    """
+    ctx = get_script_run_ctx()
+    if not ctx:
+        return None
+
+    nav_sections = {"": pages} if isinstance(pages, list) else pages
+    page_list = pages_from_nav_sections(nav_sections)
+
+    if not page_list:
+        raise StreamlitAPIException(
+            "`st.navigation` must be called with at least one `st.Page`."
+        )
+
+    default_page = None
+    pagehash_to_pageinfo: dict[PageHash, PageInfo] = {}
+
+    # This nested loop keeps track of three things:
+    # 1. the default page
+    # 2. the pagehash to pageinfo mapping
```

Comment:
nit: "keeps track of three things" but then you list only 2 😁 what's the third?

Reply:
Ah. I removed the third at the last minute (I should stop counting things 😆 )